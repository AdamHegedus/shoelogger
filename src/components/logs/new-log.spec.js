import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import BootstrapVue from 'bootstrap-vue';
import NewLog from '@/components/logs/new-log';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(BootstrapVue);

describe('new-log.vue', () => {
    describe('meta.status is false', () => {
        const logsGetters = {
            getMeta: () => {
                return {
                    status: false,
                    message: 'Error message text.'
                };
            }
        };

        const shoesActions = {
            addShoe: jest.fn()
        };

        const brandsActions = {
            getBrands: jest.fn()
        };

        const typesActions = {
            getTypes: jest.fn()
        };

        const typesState = {
            types: [
                {
                    id: 0,
                    type: 'Foo'
                }
            ]
        };

        const brandsState = {
            brands: [
                {
                    id: 0,
                    brand: 'Foo'
                }
            ]
        };

        const store = new Vuex.Store({
            modules: {
                logs: {
                    namespaced: true,
                    getters: logsGetters
                },
                shoes: {
                    namespaced: true,
                    getters: logsGetters,
                    actions: shoesActions
                },
                brands: {
                    namespaced: true,
                    state: brandsState,
                    actions: brandsActions
                },
                types: {
                    namespaced: true,
                    state: typesState,
                    actions: typesActions
                }
            }
        });

        it('should match snapshot when page loaded', () => {
            // ASSIGN
            const wrapper = shallowMount(NewLog, { store, localVue });

            // ACT
            const actual = wrapper.element;

            // ASSERT
            expect(actual).toMatchSnapshot();
        });
    });

    describe('meta.status is true', () => {
        const logsGetters = {
            getMeta: () => {
                return {
                    status: true,
                    message: null
                };
            }
        };

        const shoesActions = {
            addShoe: jest.fn()
        };

        const brandsActions = {
            getBrands: jest.fn()
        };

        const typesActions = {
            getTypes: jest.fn()
        };

        const typesState = {
            types: [
                {
                    id: 0,
                    type: 'Foo'
                }
            ]
        };

        const brandsState = {
            brands: [
                {
                    id: 0,
                    brand: 'Foo'
                }
            ]
        };

        const store = new Vuex.Store({
            modules: {
                logs: {
                    namespaced: true,
                    getters: logsGetters
                },
                shoes: {
                    namespaced: true,
                    getters: logsGetters,
                    actions: shoesActions
                },
                brands: {
                    namespaced: true,
                    state: brandsState,
                    actions: brandsActions
                },
                types: {
                    namespaced: true,
                    state: typesState,
                    actions: typesActions
                }
            }
        });

        it('should match snapshot when page loaded', () => {
            // ASSIGN
            const wrapper = shallowMount(NewLog, { store, localVue });

            // ACT
            const actual = wrapper.element;

            // ASSERT
            expect(actual).toMatchSnapshot();
        });
    });

    it('should match snapshot when distance text input has changed', () => {
        // ASSIGN
        const logsGetters = {
            getMeta: () => {
                return {
                    status: true,
                    message: null
                };
            }
        };

        const shoesActions = {
            addShoe: jest.fn()
        };

        const brandsActions = {
            getBrands: jest.fn()
        };

        const typesActions = {
            getTypes: jest.fn()
        };

        const typesState = {
            types: [
                {
                    id: 0,
                    type: 'Foo'
                }
            ]
        };

        const brandsState = {
            brands: [
                {
                    id: 0,
                    brand: 'Foo'
                }
            ]
        };

        const store = new Vuex.Store({
            modules: {
                logs: {
                    namespaced: true,
                    getters: logsGetters
                },
                shoes: {
                    namespaced: true,
                    actions: shoesActions
                },
                brands: {
                    namespaced: true,
                    state: brandsState,
                    actions: brandsActions
                },
                types: {
                    namespaced: true,
                    state: typesState,
                    actions: typesActions
                }
            }
        });

        const wrapper = shallowMount(NewLog, { store, localVue });
        const expected = '1';
        const input = '1';

        // ACT
        wrapper.find('input').setValue(input);

        // ASSERT
        const element = wrapper.element;
        const actual = wrapper.vm.$data.distance;
        expect(element).toMatchSnapshot();
        expect(actual).toEqual(expected);
    });

    describe('getShoe', () => {
        it('should return the value of text input', () => {
            // ASSIGN
            const logsGetters = {
                getMeta: () => {
                    return {
                        status: true,
                        message: null
                    };
                }
            };

            const shoesActions = {
                addShoe: jest.fn()
            };

            const brandsActions = {
                getBrands: jest.fn()
            };

            const typesActions = {
                getTypes: jest.fn()
            };

            const typesState = {
                types: [
                    {
                        id: 0,
                        type: 'Foo'
                    }
                ]
            };

            const brandsState = {
                brands: [
                    {
                        id: 0,
                        brand: 'Foo'
                    }
                ]
            };

            const store = new Vuex.Store({
                modules: {
                    logs: {
                        namespaced: true,
                        getters: logsGetters
                    },
                    shoes: {
                        namespaced: true,
                        getters: logsGetters,
                        actions: shoesActions
                    },
                    brands: {
                        namespaced: true,
                        state: brandsState,
                        actions: brandsActions
                    },
                    types: {
                        namespaced: true,
                        state: typesState,
                        actions: typesActions
                    }
                }
            });

            const wrapper = shallowMount(NewLog, { store, localVue });
            const input = '1';
            wrapper.vm.$data.selectedShoe = {
                id: 0,
                brandId: 0,
                typeId: 0,
                product: 'Foo'
            };
            wrapper.vm.$data.distance = 10;
            wrapper.vm.$data.date = new Date('2000-01-01').toISOString().substring(0, 10);
            wrapper.find('#distance').setValue(input);

            // ACT
            const actual = wrapper.vm.getShoe;

            // ASSERT
            expect(actual).toMatchSnapshot();
        });
    });

    describe('shoes', () => {
        it('should return the value from store', () => {
            // ASSIGN
            const logsGetters = {
                getMeta: () => {
                    return {
                        status: true,
                        message: null
                    };
                }
            };

            const shoesState = {
                shoes: [
                    {
                        id: 0,
                        brandId: 0,
                        typeId: 0,
                        product: 'Foo'
                    },
                    {
                        id: 1,
                        brandId: 0,
                        typeId: 0,
                        product: 'Bar'
                    }
                ]
            };

            const typesState = {
                types: [
                    {
                        id: 0,
                        type: 'Foo'
                    }
                ]
            };

            const brandsState = {
                brands: [
                    {
                        id: 0,
                        brand: 'Foo'
                    }
                ]
            };

            const store = new Vuex.Store({
                modules: {
                    logs: {
                        namespaced: true,
                        getters: logsGetters
                    },
                    shoes: {
                        namespaced: true,
                        state: shoesState
                    },
                    brands: {
                        namespaced: true,
                        state: brandsState
                    },
                    types: {
                        namespaced: true,
                        state: typesState
                    }
                }
            });

            const wrapper = shallowMount(NewLog, { store, localVue });

            // ACT
            const actual = wrapper.vm.shoes;

            // ASSERT
            expect(actual).toMatchSnapshot();
        });
    });

    describe('addLog', () => {
        it('should call action', () => {
            // ASSIGN
            const logsGetters = {
                getMeta: () => {
                    return {
                        status: true,
                        message: null
                    };
                }
            };

            const logsActions = {
                addLog: jest.fn()
            };

            const shoesState = {
                shoes: [
                    {
                        id: 0,
                        brandId: 0,
                        typeId: 0,
                        product: 'Foo'
                    },
                    {
                        id: 1,
                        brandId: 0,
                        typeId: 0,
                        product: 'Bar'
                    }
                ]
            };

            const brandsActions = {
                getBrands: jest.fn()
            };

            const typesActions = {
                getTypes: jest.fn()
            };

            const typesState = {
                types: [
                    {
                        id: 1,
                        type: 'Bar'
                    }
                ]
            };

            const brandsState = {
                brands: [
                    {
                        id: 0,
                        brand: 'Foo'
                    }
                ]
            };

            const store = new Vuex.Store({
                modules: {
                    logs: {
                        namespaced: true,
                        actions: logsActions,
                        getters: logsGetters
                    },
                    shoes: {
                        namespaced: true,
                        state: shoesState,
                        getters: logsGetters
                    },
                    brands: {
                        namespaced: true,
                        state: brandsState,
                        actions: brandsActions
                    },
                    types: {
                        namespaced: true,
                        state: typesState,
                        actions: typesActions
                    }
                }
            });

            const wrapper = shallowMount(NewLog, { store, localVue });
            const input = {
                shoeId: 0,
                distance: 10,
                date: 1000000000
            };

            // ACT
            wrapper.vm.addLog(input);

            // ASSERT
            const actual = logsActions.addLog.mock.calls;
            expect(actual).toMatchSnapshot();
        });
    });

    describe('getShoeDisplayValue', () => {
        it('should return correct value', () => {
            // ASSIGN
            const logsGetters = {
                getMeta: () => {
                    return {
                        status: true,
                        message: null
                    };
                }
            };

            const shoesActions = {
                addShoe: jest.fn()
            };

            const brandsActions = {
                getBrands: jest.fn()
            };

            const typesActions = {
                getTypes: jest.fn()
            };

            const typesState = {
                types: [
                    {
                        id: 1,
                        type: 'Bar'
                    }
                ]
            };

            const brandsState = {
                brands: [
                    {
                        id: 0,
                        brand: 'Foo'
                    }
                ]
            };

            const store = new Vuex.Store({
                modules: {
                    logs: {
                        namespaced: true,
                        getters: logsGetters
                    },
                    shoes: {
                        namespaced: true,
                        getters: logsGetters,
                        actions: shoesActions
                    },
                    brands: {
                        namespaced: true,
                        state: brandsState,
                        actions: brandsActions
                    },
                    types: {
                        namespaced: true,
                        state: typesState,
                        actions: typesActions
                    }
                }
            });

            const wrapper = shallowMount(NewLog, { store, localVue });
            const input = {
                id: 1,
                brand: 'Bar',
                product: 'Foo'
            };

            // ACT
            const actual = wrapper.vm.getShoeDisplayValue(input);

            // ASSERT
            expect(actual).toMatchSnapshot();
        });
    });
});
