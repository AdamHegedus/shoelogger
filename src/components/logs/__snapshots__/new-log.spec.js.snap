// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new-log.vue addLog should call action 1`] = `
Array [
  Array [
    Object {
      "commit": [Function],
      "dispatch": [Function],
      "getters": Object {
        "getMeta": Object {
          "message": null,
          "status": true,
        },
      },
      "rootGetters": Object {
        "logs/getMeta": Object {
          "message": null,
          "status": true,
        },
        "shoes/getMeta": Object {
          "message": null,
          "status": true,
        },
      },
      "rootState": Object {
        "brands": Object {
          "brands": Array [
            Object {
              "brand": "Foo",
              "id": 0,
            },
          ],
        },
        "logs": Object {},
        "shoes": Object {
          "shoes": Array [
            Object {
              "brandId": 0,
              "id": 0,
              "product": "Foo",
              "typeId": 0,
            },
            Object {
              "brandId": 0,
              "id": 1,
              "product": "Bar",
              "typeId": 0,
            },
          ],
        },
        "types": Object {
          "types": Array [
            Object {
              "id": 1,
              "type": "Bar",
            },
          ],
        },
      },
      "state": Object {},
    },
    Object {
      "date": 1000000000,
      "distance": 10,
      "shoeId": 0,
    },
    undefined,
  ],
]
`;

exports[`new-log.vue getShoe should return the value of text input 1`] = `
Object {
  "date": "2000-01-01",
  "distance": "1",
  "shoeId": 0,
}
`;

exports[`new-log.vue meta.status is false should match snapshot when page loaded 1`] = `
<div
  class="container"
>
  <div
    class="row"
  >
    <div
      class="col-sm col-lg-6"
    >
      <h2>
        Add New Log Entry
      </h2>
       
      <form>
        <div
          class="form-group"
        >
          <div
            class="form-group"
          >
            <label>
              Shoe
            </label>
             
            <b-dropdown-stub
              boundary="scrollParent"
              class="btn-block"
              id="type"
              no-flip="true"
              role="menu"
              size="lg"
              text="Select Shoe"
              toggleclass="btn btn-secondary btn-lg btn-block shadow"
              toggletext="Toggle Dropdown"
            />
          </div>
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="distance"
          >
            Distance
          </label>
           
          <input
            class="form-control form-control-lg shadow border border-primary"
            id="distance"
            placeholder="Distance (km)"
            type="number"
          />
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="date"
          >
            Date
          </label>
           
          <input
            class="form-control form-control-lg shadow border border-primary"
            id="date"
            placeholder="Date (YYYY-mm-dd)"
            type="date"
          />
        </div>
         
        <button
          class="btn btn-primary btn-lg btn-block shadow"
          disabled="disabled"
          type="submit"
        >
          
                    Add
                
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`new-log.vue meta.status is true should match snapshot when page loaded 1`] = `
<div
  class="container"
>
  <div
    class="row"
  >
    <div
      class="col-sm col-lg-6"
    >
      <h2>
        Add New Log Entry
      </h2>
       
      <form>
        <div
          class="form-group"
        >
          <div
            class="form-group"
          >
            <label>
              Shoe
            </label>
             
            <b-dropdown-stub
              boundary="scrollParent"
              class="btn-block"
              id="type"
              no-flip="true"
              role="menu"
              size="lg"
              text="Select Shoe"
              toggleclass="btn btn-secondary btn-lg btn-block shadow"
              toggletext="Toggle Dropdown"
            />
          </div>
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="distance"
          >
            Distance
          </label>
           
          <input
            class="form-control form-control-lg shadow border border-primary"
            id="distance"
            placeholder="Distance (km)"
            type="number"
          />
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="date"
          >
            Date
          </label>
           
          <input
            class="form-control form-control-lg shadow border border-primary"
            id="date"
            placeholder="Date (YYYY-mm-dd)"
            type="date"
          />
        </div>
         
        <button
          class="btn btn-primary btn-lg btn-block shadow"
          disabled="disabled"
          type="submit"
        >
          
                    Add
                
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`new-log.vue selectShoe should set selectedShoe 1`] = `
Object {
  "brandId": 0,
  "id": 1,
  "product": "Foo",
  "typeId": 0,
}
`;

exports[`new-log.vue shoes should return the value from store 1`] = `undefined`;

exports[`new-log.vue should match snapshot when distance text input has changed 1`] = `
<div
  class="container"
>
  <div
    class="row"
  >
    <div
      class="col-sm col-lg-6"
    >
      <h2>
        Add New Log Entry
      </h2>
       
      <form>
        <div
          class="form-group"
        >
          <div
            class="form-group"
          >
            <label>
              Shoe
            </label>
             
            <b-dropdown-stub
              boundary="scrollParent"
              class="btn-block"
              id="type"
              no-flip="true"
              role="menu"
              size="lg"
              text="Select Shoe"
              toggleclass="btn btn-secondary btn-lg btn-block shadow"
              toggletext="Toggle Dropdown"
            />
          </div>
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="distance"
          >
            Distance
          </label>
           
          <input
            class="form-control form-control-lg shadow border border-primary"
            id="distance"
            placeholder="Distance (km)"
            type="number"
          />
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="date"
          >
            Date
          </label>
           
          <input
            class="form-control form-control-lg shadow border border-primary"
            id="date"
            placeholder="Date (YYYY-mm-dd)"
            type="date"
          />
        </div>
         
        <button
          class="btn btn-primary btn-lg btn-block shadow"
          disabled="disabled"
          type="submit"
        >
          
                    Add
                
        </button>
      </form>
    </div>
  </div>
</div>
`;
